noir-projects/noir-protocol-circuits/crates/rollup-lib/src/components.nr

```diff
@@ -275,6 +275,226 @@ pub fn append_tx_effects_for_blob(
     l2_to_l1_msgs: [Field; MAX_L2_TO_L1_MSGS_PER_TX],
     start_sponge_blob: SpongeBlob,
 ) -> SpongeBlob {
+    let (mut tx_effects_hash_input, offset) = get_tx_effects_hash_input(
+        combined,
+        revert_code,
+        transaction_fee,
+        all_public_data_update_requests,
+        l2_to_l1_msgs,
+    );
+
+    // NB: using start.absorb & returning start caused issues in ghost values appearing in
+    // base_rollup_inputs.start when using a fresh sponge. These only appeared when simulating via wasm.
+    let mut out_sponge = start_sponge_blob;
+
+    // If we have an empty tx (usually a padding tx), we don't want to absorb anything
+    // An empty tx will only have 2 effects - revert code and fee - hence offset = 2
+    if offset != 2 {
+        out_sponge.absorb(tx_effects_hash_input, offset);
+    }
+
+    out_sponge
+}
+
+fn get_tx_effects_hash_input(
+    combined: CombinedAccumulatedData,
+    revert_code: u8,
+    transaction_fee: Field,
+    all_public_data_update_requests: [PublicDataWrite; MAX_TOTAL_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX],
+    l2_to_l1_msgs: [Field; MAX_L2_TO_L1_MSGS_PER_TX],
+) -> ([Field; TX_EFFECTS_BLOB_HASH_INPUT_FIELDS], u32) {
+    let mut tx_effects_hash_input = unsafe {
+        get_tx_effects_hash_input_helper(
+            combined,
+            revert_code,
+            transaction_fee,
+            all_public_data_update_requests,
+            l2_to_l1_msgs,
+        )
+    };
+
+    let note_hashes = combined.note_hashes;
+    let nullifiers = combined.nullifiers;
+
+    // Public writes are the concatenation of all non-empty user update requests and protocol update requests, then padded with zeroes.
+    // The incoming all_public_data_update_requests may have empty update requests in the middle, so we move those to the end of the array.
+    let public_data_update_requests =
+        get_all_update_requests_for_tx_effects(all_public_data_update_requests);
+    let private_logs = combined.private_logs;
+    let unencrypted_logs =
+        combined.unencrypted_logs_hashes.map(|log: ScopedLogHash| silo_unencrypted_log_hash(log));
+    let contract_class_logs = combined.contract_class_logs_hashes.map(|log: ScopedLogHash| {
+        silo_unencrypted_log_hash(log)
+    });
```

Comment:
I'd recommend refactoring the helper function a bit so we're passing in these arrays after they're processed in constrained-land as we can shrink the amount of brillig which needs to be executed (plus generally avoid copy-pasta code like we have now)

Reply:
Sounds good, I'd be happy to refactor this. So for example, gather the `public_data_update_requests` in constrained at the beginning of this fn then pass the result to `get_tx_effects_hash_input_helper`?

Reply:
Yep, that's it.

Reply:
https://github.com/AztecProtocol/aztec-packages/issues/11058